[B] Added Location.to{World,Local}{Axis,}(Vector). Adds BUKKIT-5338

TomyLobo opened the pull request:
#### The Issue

The Location class currently lacks methods for transforming between coordinate systems easily.
#### Justification for this PR

This PR allows developers to do the affine transformation to and from a local coordinate system that rotates and moves with the reference Location, whereby new Location(*, 0, 0, 0, 0, 0) does no transformation at all.

This implies that:
- Positive y is up
- Positive z (south, i.e. yaw=0) is forward
- Positive x (east, which is left of south) is left
- getDirection() is equivalent to toWorldAxis(new Vector(0,0,1))

This can be used to render effects adjacent to the player or creature and to spawn projectiles in a consistent pattern relative to the player's view.
#### PR Breakdown

The toWorld\* methods transform from local coordinates to world coordinates.
The toLocal\* methods transform from world coordinates to local coordinates.
The *Axis methods deal with axes instead of positions.
#### Testing Results and Materials

Unit tests are included and should be sufficient to ensure that the new methods are working.
#### Relevant PR(s)

[B-1012](https://github.com/Bukkit/Bukkit/issues/1012)
#### JIRA Ticket

[BUKKIT-5338](https://bukkit.atlassian.net/browse/BUKKIT-5338)

TomyLobo committed:
Added Location.to{World,Local}{Axis,}(Vector).

The methods transform the specified axis or position according to the reference
frame given by the Location.

When called on new Location(*, 0, 0, 0, 0, 0), the methods do nothing.

This implies that:
- Positive x (east, left of yaw=0) is left
- Positive y is up
- Positive z (south, yaw=0) is forward

Furthermore, toLocal(toWorld(x)) returns x, same with the function
invocations swapped and same with axis methods instead of position
methods.

turt2live commented:
I can honestly say I have no idea what the math involved here is doing/should do, but I'll look over the formatting in a moment.

In the meantime, please update your title to match the CONTRIBUTING.md guidelines.

turt2live commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,106 @@ public Location clone() {
+     * Transforms an axis from the local reference frame described by this
+     * location into the world reference frame.
+     * <p>
+     * x=left, y=up, z=forward
```
These 3 variables should be in sentence form or list form with an applicable sentence. 

Such as `.. where x represents the left...`

or `... where the following conditions are applied: <list>`

TomyLobo replied:
I'll use the later. thanks for the feedback.

turt2live commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,106 @@ public Location clone() {
+        final double cosPitch = Math.cos(pitch);
+        final double sinPitch = Math.sin(pitch);
+
+        final Vector forward = new Vector(
```
This does not need to be spread over 5 lines, nor do the following vectors.

Additionally, `x * x` should be used as a format, not `x*x` (note the spacing)

TomyLobo replied:
fixed locally

turt2live commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,106 @@ public Location clone() {
+        final double cosPitch = Math.cos(pitch);
+        final double sinPitch = Math.sin(pitch);
+
+        final Vector xAxis = new Vector(
```
https://github.com/Bukkit/Bukkit/pull/1012/files#r9171780

TomyLobo replied:
fixed locally

turt2live commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,106 @@ public Location clone() {
+     * Transforms an axis from the world reference frame into the local reference frame described by this
+     * location.
+     * <p>
+     * x=left, y=up, z=forward
```
https://github.com/Bukkit/Bukkit/pull/1012/files#r9171772

turt2live commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,106 @@ public Location clone() {
+     * Transforms a position from the local reference frame described by this
+     * location into the world reference frame.
+     * <p>
+     * x=left, y=up, z=forward
```
https://github.com/Bukkit/Bukkit/pull/1012/files#r9171772

turt2live commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,106 @@ public Location clone() {
+     * Transforms a position from the world reference frame into the local reference frame described by this
+     * location.
+     * <p>
+     * x=left, y=up, z=forward
```
https://github.com/Bukkit/Bukkit/pull/1012/files#r9171772

TomyLobo committed:
fixup! Changes to improve code and documentation quality.

TomyLobo commented:
I pushed changes addressing your comments.
the maths:
Basically this is 2 parametric rotation matrices (one for yaw, one for pitch), multiplied into one in a specific order.
in toWorldAxis (582-598), you see the matrix split up into 3 vectors.
Since rotation matrices are orthonormal, you can invert them by simply transposing them.
You see that inverted matrix in toLocalAxis (624-640), again split up into 3 vectors.

Each of these 3 vectors is one of the base axes of the target coordinate system, represented in the source coordinate system.
Multiplying them by the components corresponding to those axes and adding it all up yields the transformed axis.

Wolvereness commented on src/main/java/org/bukkit/Location.java:
```diff
@@ -557,4 +557,120 @@ public Location clone() {
+     * location into the world reference frame.
+     * <p>
+     * The following conditions apply:
+     * <li>X axis points to the left</li>
```
We don't use closing list item markers, and this needs a unordered list market before it starts.

Wolvereness commented:
I normally take pride in my knowledge of mathematical concepts, but I have no idea what this is supposed to do. Intended input and output, with use case isn't documented very well. Based on what little description you gave, this is already possible using the current api, without hand-rolling the math.

I also want to see unit tests parameterized with expected inputs and outputs, like https://github.com/Bukkit/Bukkit/blob/master/src/test/java/org/bukkit/LocationTest.java

Oh, and if you have to assign an arbitrary direction as 'forward' , I'm pretty sure it's not appropriate api.

TomyLobo committed:
fixup! Copied list format from org.bukkit.Location.setYaw.

TomyLobo commented:
> I normally take pride in my knowledge of mathematical concepts, but I have no idea what this is supposed to do.

I explained it in [my comment](https://github.com/Bukkit/Bukkit/pull/1012#issuecomment-33291571).

> Intended input and output, with use case isn't documented very well.

Described in the justification and by the test cases.
I'll translate the test cases into plain english and attach them to the javadocs.

> Based on what little description you gave, this is already possible using the current api, without hand-rolling the math.

Can you explain how?
EDIT: I did a "find usages" on Math.cos and found nothing outside Location.java and only getDirection inside Location.java.
And seeing as a cosine would be necessary for this, I'm going to disagree here and claim that there is _no_ functionality like this in Bukkit yet.

> I also want to see unit tests parameterized with expected inputs and outputs, like https://github.com/Bukkit/Bukkit/blob/master/src/test/java/org/bukkit/LocationTest.java

I don't know how these parameterized tests work, but I did see that test you made. I roughly see what's going on, but seeing as I had all of my tests parametrized already, I don't see how it's necessary to make them conform to that framework you're using that I'm not familiar with.

> Oh, and if you have to assign an arbitrary direction as 'forward' , I'm pretty sure it's not appropriate api.

I agree with that statement.
However, if you face direction 0 (south) and then walk forward, you'll notice z goes up, so assigning positive z to forward isn't arbitrary.
This kind of conformity is covered by a unit test, too. I'd be happy to explain these to you if you ping me on IRC.
The gist of that conversation can then become a comment or javadoc or something.

_(edited the last time at 10:08 CET, 16 minutes after posting)_

TomyLobo commented:
thought of a better solution

Wolvereness commented:
Context: https://gist.github.com/Wolvereness/8631025

TomyLobo commented:
Status: After messing around with the matrix multiplication order and angle units for a bit, I made it conform to the tests I used for the previous proposal.
That implies, however, that I'm not testing roll at all and there's nothing in the API that uses roll which I could ensure conformity to with another test.
That test framework where you specify inputs and outputs would be useful here :)

Wolvereness commented:
What happened to this?

TomyLobo commented:
On #bukkitdev, I defended a guy who got banned for using "stuff.com" (a malware site) as a placeholder.
I shouldn't have been surprised, but that got me banned as well.

I think that's not a way to treat people volunteering their time, so I'm no longer doing that.
I explained it on the ticket, back then.
